engines/wage/design.cpp

```diff
@@ -0,0 +1,534 @@
+		in.readByte();
+
+	FloodFill ff(&tmp, kColorWhite, kColorGreen);
+	for (int yy = 0; yy < h; yy++) {
+		ff.addSeed(0, yy);
+		ff.addSeed(w - 1, yy);
+	}
+	for (int xx = 0; xx < w; xx++) {
+		ff.addSeed(xx, 0);
+		ff.addSeed(xx, h - 1);
+	}
+	ff.fill();
+
+	for (y = 0; y < h; y++) {
+		byte *src = (byte *)tmp.getBasePtr(0, y);
+		byte *dst = (byte *)surface->getBasePtr(x1, y1 + y);
+		for (x = 0; x < w; x++) {
+			if (*src != kColorGreen)
+				*dst = *src;
+			src++;
+			dst++;
+		}
+	}
+
+	tmp.free();
+}
+
+void Design::drawRect(Graphics::Surface *surface, Common::Rect &rect, int thickness, int color, Patterns &patterns, byte fillType) {
+	drawRect(surface, rect.left, rect.top, rect.right, rect.bottom, thickness, color, patterns, fillType);
+}
+
+void Design::drawRect(Graphics::Surface *surface, int x1, int y1, int x2, int y2, int thickness, int color, Patterns &patterns, byte fillType) {
+	plotData pd(surface, &patterns, fillType, thickness);
+
+	Graphics::drawLine(x1, y1, x2, y1, kColorBlack, drawPixel, &pd);
+	Graphics::drawLine(x2, y1, x2, y2, kColorBlack, drawPixel, &pd);
+	Graphics::drawLine(x2, y2, x1, y2, kColorBlack, drawPixel, &pd);
+	Graphics::drawLine(x1, y2, x1, y1, kColorBlack, drawPixel, &pd);
+}
+
+
+void Design::drawFilledRect(Graphics::Surface *surface, Common::Rect &rect, int color, Patterns &patterns, byte fillType) {
+	plotData pd(surface, &patterns, fillType, 1);
+
+	for (int y = rect.top; y <= rect.bottom; y++)
+		Graphics::drawHLine(rect.left, rect.right, y, color, drawPixel, &pd);
+}
+
+void Design::drawFilledRoundRect(Graphics::Surface *surface, Common::Rect &rect, int arc, int color, Patterns &patterns, byte fillType) {
+	plotData pd(surface, &patterns, fillType, 1);
+
+	Graphics::drawRoundRect(rect, arc, color, true, drawPixel, &pd);
+}
+
+void Design::drawHLine(Graphics::Surface *surface, int x1, int x2, int y, int thickness, int color, Patterns &patterns, byte fillType) {
+	plotData pd(surface, &patterns, fillType, thickness);
+
+	Graphics::drawHLine(x1, x2, y, color, drawPixel, &pd);
+}
+
+void Design::drawVLine(Graphics::Surface *surface, int x, int y1, int y2, int thickness, int color, Patterns &patterns, byte fillType) {
+	plotData pd(surface, &patterns, fillType, thickness);
+
+	Graphics::drawVLine(x, y1, y2, color, drawPixel, &pd);
+}
+
+FloodFill::FloodFill(Graphics::Surface *surface, byte color1, byte color2) {
+	_surface = surface;
+	_color1 = color1;
+	_color2 = color2;
+	_w = surface->w;
+	_h = surface->h;
+
+	_visited = (byte *)calloc(_w * _h, 1);
+}
+
+FloodFill::~FloodFill() {
+	while(_queue.size()) {
+		Common::Point *p = _queue.front();
+
+		delete p;
+		_queue.pop_front();
+	}
+
+	free(_visited);
+}
+
+void FloodFill::addSeed(int x, int y) {
+	byte *p;
+
+	if (x >= 0 && x < _w && y >= 0 && y < _h) {
+		if (!_visited[y * _w + x] && *(p = (byte *)_surface->getBasePtr(x, y)) == _color1) {
+			_visited[y * _w + x] = 1;
+			*p = _color2;
+
+			Common::Point *pt = new Common::Point(x, y);
```

Comment:
I don't think using a List of Point pointers makes sense here compared to simply using a List of Point objects. (The object is only 4 bytes.)